File: frontend/lib/types/exchange.ts
export type ExchangeStatus =
  | 'pending'
  | 'accepted'
  | 'declined'
  | 'cancelled'
  | 'invalid'
  | 'expired';

export interface ShiftAssignment {
  date: string;
  day_of_week: string;
  employee_name: string;
}

export interface ExchangeRequest {
  id: number;
  month_year: string;
  requester_employee_id: number;
  requester_employee_name: string;
  requester_date: string;
  target_employee_id: number;
  target_employee_name: string;
  target_date: string;
  status: ExchangeStatus;
  reason?: string;
  decline_reason?: string;
  validation_errors?: string[];
  created_at: string;
  responded_at?: string;
  completed_at?: string;
}

export interface SwapCandidate {
  employee_id: number;
  employee_name: string;
  eligible_dates: string[];
}

export interface ShiftWithCalendarLink {
  date: string;
  day_of_week: string;
  calendar_url: string;
}

export interface ExchangeNotification {
  id: string;
  type: string;
  message: string;
  exchange?: ExchangeRequest;
  shifts?: ShiftWithCalendarLink[];
  timestamp: string;
  read: boolean;
}

export interface WSMessage {
  type: string;
  exchange?: ExchangeRequest;
  shifts?: ShiftWithCalendarLink[];
  message: string;
  month_year?: string;
}

// Schedule calendar types
export interface ScheduleDateCell {
  date: string; // YYYY-MM-DD
  dayNumber: number;
  assignedEmployee: string | null;
  assignedEmployeeId: number | null;
  isCurrentUserShift: boolean;
  isPast: boolean;
  isWeekend: boolean; // Friday or Saturday
  hasPendingExchange: boolean;
}

export interface MonthSchedule {
  year: number;
  month: number;
  firstDayOffset: number;
  daysInMonth: number;
  dates: ScheduleDateCell[];
  currentUserShiftDates: string[];
}

export interface EmployeeAvailability {
  employeeId: number;
  name: string;
  availableDates: string[];
  unavailableDates: string[];
}

export interface EnhancedSwapCandidate extends SwapCandidate {
  is_new: boolean;
  all_shift_dates: string[];
  availability?: EmployeeAvailability;
}

File: backend/app/schemas/exchange.py
"""Shift exchange schemas for swap requests and responses."""

from enum import Enum
from typing import Optional
from pydantic import BaseModel


class ExchangeStatus(str, Enum):
    """Exchange request status."""
    PENDING = "pending"
    ACCEPTED = "accepted"
    DECLINED = "declined"
    CANCELLED = "cancelled"
    INVALID = "invalid"
    EXPIRED = "expired"


class ShiftExchangeCreate(BaseModel):
    """Request to create a shift exchange."""
    requester_date: str  # YYYY-MM-DD
    target_employee_id: int
    target_date: str  # YYYY-MM-DD
    reason: Optional[str] = None


class ShiftExchangeResponse(BaseModel):
    """Response model for a shift exchange."""
    id: int
    month_year: str
    requester_employee_id: int
    requester_employee_name: str
    requester_date: str
    requester_shift_type: Optional[str] = None
    target_employee_id: int
    target_employee_name: str
    target_date: str
    target_shift_type: Optional[str] = None
    status: ExchangeStatus
    reason: Optional[str] = None
    decline_reason: Optional[str] = None
    validation_errors: Optional[list[str]] = None
    created_at: str
    responded_at: Optional[str] = None
    completed_at: Optional[str] = None


class ShiftExchangeAction(BaseModel):
    """Action on an exchange request."""
    action: str  # "accept" or "decline"
    decline_reason: Optional[str] = None


class EligibleSwapPartner(BaseModel):
    """An eligible partner for a shift swap."""
    employee_id: int
    employee_name: str
    eligible_dates: list[str]
